go/merkletree2/tree.go

```diff
@@ -1,38 +1,405 @@
+// Internal nodes store the hashes of their children, while leaves store store
+// KeyValuePairsHashed (the Value in each pair is not required to verify the
+// hash of a a leaf node).
+type Node struct {
+	INodes     []Hash        `codec:"i,omitempty"`
+	LeafHashes []KeyHashPair `codec:"l,omitempty"`
+	Type       NodeType      `codec:"t"`
+}
+
+type PositionHashPair struct {
+	p Position
+	h Hash
+}
+
+type RootMetadataNode struct {
+	// TODO Add skip pointers, timestamp, version.....
+	Seqno        Seqno
+	BareRootHash Hash
+}
+
+func (t *Tree) makeNextRootMNode(curr *RootMetadataNode, newRootHash Hash) RootMetadataNode {
+	if curr == nil {
+		return RootMetadataNode{Seqno: 1, BareRootHash: newRootHash}
+	}
+	return RootMetadataNode{Seqno: curr.Seqno + 1, BareRootHash: newRootHash}
+}
+
+func (t *Tree) GenerateAndStoreMasterSecret(
+	ctx context.Context, tr Transaction, s Seqno) (ms MasterSecret, err error) {
+	ms, err = t.hasher.GenerateMasterSecret(s)
+	if err != nil {
+		return nil, err
+	}
+	err = t.eng.StoreMasterSecret(ctx, tr, s, ms)
+	if err != nil {
+		return nil, err
+	}
+	return ms, nil
+}
+
+// BuildNewTreeVersionFromAllKeys builds a new tree version, taking a batch
+// input. The sortedKeyValuePairs must be sorted (lexicographically) by Key.
+//
+// NOTE: The input to this function should contain at least all the keys which
+// were inserted in previous versions of the tree, otherwise this procedure will
+// put the tree into an inconsistent state. This function does not check the
+// condition is true for efficiency reasons.
+func (t *Tree) BuildNewTreeVersionFromAllKeys(
+	ctx context.Context, tr Transaction, sortedKVPairs []KeyValuePair) (err error) {
+	t.Lock()
+	defer t.Unlock()
+
+	var latestSeqNo Seqno
+	var rootMNode RootMetadataNode
+	if latestSeqNo, rootMNode, err = t.eng.LookupLatestRoot(ctx, tr); err != nil {
+		return err
+	}
+	if latestSeqNo == 0 {
+		t.log.CWarningf(ctx, "No root found. Starting a new merkle tree.")
+	}
+
+	var ms MasterSecret
+	if t.useBlindedValueHashes {
+		ms, err = t.GenerateAndStoreMasterSecret(ctx, tr, latestSeqNo+1)
+		if err != nil {
+			return err
+		}
+	} else {
+		ms = nil
+	}
+
+	var newRootHash Hash
+	if newRootHash, err = t.hashTreeRecursive(ctx, tr, latestSeqNo+1, ms,
+		*t.getRootPosition(), sortedKVPairs); err != nil {
+		return err
+	}
+
+	newRoot := t.makeNextRootMNode(&rootMNode, newRootHash)
+
+	if err = t.eng.StoreRootMetadataNode(ctx, tr, newRoot); err != nil {
+		return err
+	}
+
+	if err = t.eng.CommitTransaction(ctx, tr); err != nil {
+		return err
+	}
+
+	return err
+}
+
+func (t *Tree) hashTreeRecursive(ctx context.Context, tr Transaction, s Seqno, ms MasterSecret,
+	p Position, sortedKVPairs []KeyValuePair) (ret Hash, err error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	if len(sortedKVPairs) <= t.maxValuesPerLeaf {
+		ret, err = t.makeAndStoreLeaf(ctx, tr, s, ms, p, sortedKVPairs)
+		return ret, err
+	}
+
+	var node Node
```

Comment:
do node := Node{Type: ... INodes: ...}
or better yet, make two constructors like NewEmptyInternalNode and NewLeafNode